### PR TITLE
[MRG] refactoring of external feed class

### PR DESCRIPTION
**Pull request by jasmainak**

### CODE REVIEW

**Comment by jasmainak** [comment]
yes go for it! Let me know when you're done

**Comment by rythorpe** [comment]
> yes go for it! Let me know when you're done

Done for now! I still need to update the corresponding params but I'll need to save that for later.

**Comment by jasmainak** [comment]
Great!

Btw, do you know if the GUI has a provision for 'normal' vs 'uniform' distribution. I am a little puzzled to [see it in the code](https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/feed.py#L246) but not in the GUI?

**Comment by rythorpe** [comment]
> Great!
> 
> Btw, do you know if the GUI has a provision for 'normal' vs 'uniform' distribution. I am a little puzzled to [see it in the code](https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/feed.py#L246) but not in the GUI?

You're right, I don't see the 'uniform' burst distribution referenced anywhere in the GUI or mentioned in the tutorials. I'm guessing this is due to lack of a use-case. Physiologically speaking, I can't think of a situation where a burst event would ever occur with sharp edges. Even a near-synchronous Poisson process wouldn't contain the edge effects present at the bounds of a uniform (i.e., rectangular) distribution.

**Comment by jasmainak** [comment]
okay sounds good. I ask this because it means we don't have to expose an argument called `distribution` (at least for now).

**Comment by jasmainak** [comment]
@rythorpe I'm going to do a pass now. Do you have any commits to push?

**Comment by rythorpe** [comment]
Go for it. I won’t be able to work on it until Monday probably.

File: hnn_core/network.py [hunk 1]
```diff
@@ -124,9 +125,7 @@ def __init__(self, params, n_jobs=1):
         # Global number of external inputs ... automatic counting
         # makes more sense
         # p_unique represent ext inputs that are going to go to each cell
-        self.p_common, self.p_unique = create_pext(self.params,
-                                                   self.params['tstop'])
```
Comment (jasmainak) [inline hnn_core/network.py:128]: most of what is in this function needs to go inside `Rhythmic` and `Evoked` for backwards compatibility. With the new objects, users should be able to either provide legacy `params` to pre-populate these objects, or set them manually.

File: hnn_core/pyramidal.py [hunk 2]
```diff
@@ -872,24 +872,35 @@ def parconnect(self, gid, gid_dict, pos_dict, p):
                       'L2_basket', 'L2Basket', lamtha=50.,
                       postsyns=[self.apicaltuft_gabaa])
 
-    # receive from common inputs
     # XXX check NetCon connections for proximal inputs with zero weights
-    def parreceive(self, gid, gid_dict, pos_dict, p_ext):
-        for gid_src, p_src, pos in zip(gid_dict['common'],
-                                       p_ext, pos_dict['common']):
+    def parreceive(self, gids, pos_dict, feeds, threshold):
```
Comment (jasmainak) [inline hnn_core/pyramidal.py:876]: `gid` was not used anywhere in the function and `gid_dict` is very opaque. All it does is provide a range of `gids`. This should be passed directly.

File: hnn_core/pyramidal.py [hunk 3]
```diff
@@ -872,24 +872,35 @@ def parconnect(self, gid, gid_dict, pos_dict, p):
...
+        """Receive from common inputs.
+
+        Parameters
+        ----------
+        gids : dict
```
Comment (jasmainak) [inline hnn_core/pyramidal.py:881]: oops, should be a list

File: hnn_core/feed.py [hunk 4]
```diff
@@ -7,6 +7,228 @@
...
+                g_basket_nmda=params['input_dist_A_weight_L2Basket_nmda'],
+                delay=params['input_dist_A_delay_L5'])
+            self.set_dist_connections_layer5(
+                g_pyr_ampa=params['input_dist_A_weight_L5Pyr_ampa'],
+                g_pyr_nmda=params['input_dist_A_weight_L5Pyr_nmda'],
+                delay=params['input_dist_A_delay_L2'])
+
+    def set_times(self, tstart, tstop, tstart_dev, burst_freq=10, burst_std=20,
+                  burst_n_spikes=2, n_bursts=10, randomize_tstart=False,
+                  random_state1=None, random_state2=None):
```
Comment (jasmainak) [inline hnn_core/feed.py:68]: simplified version of the existing function.

File: hnn_core/feed.py [hunk 5]
```diff
@@ -7,6 +7,228 @@
...
+        self.L2_Pyr_ampa['lamtha'] = 100.
+
+        self.L2_Pyr_nmda['g'] = g_pyr_nmda
+        self.L2_Pyr_nmda['delay'] = delay
+        self.L2_Pyr_ampa['lamtha'] = 100.
+
+        self.L2_Basket_ampa['g'] = g_basket_ampa
+        self.L2_Basket_ampa['g'] = delay
+        self.L2_Basket_ampa['lamtha'] = 100.
+        pass
```
Comment (jasmainak) [inline hnn_core/feed.py:162]:
```suggestion
```

**Comment by jasmainak** [comment]
@rythorpe I have to stop for today and I may not be able to get to this tomorrow.

I left some comments. Hopefully it begins to make sense. What we're trying to do is:

1) discourage users from editing the params object directly but rather through an object oriented interface. This allows you to have proper checks and do things more transparently
2) cleaning some code along the way (getting rid of dead code for instance) + adding tests
3) documenting the parameters in the docstrings

File: hnn_core/feed.py [hunk 6]
```diff
@@ -7,6 +7,228 @@
...
+        self.loc = loc
+        self.t_input = None
+
+        if params is not None:
+            self.set_times(tstart=params[f't0_input_{loc}'],
+                           tstop=params[f'tstop_input_{loc}'],
+                           burst_freq=params[f'f_input_{loc}'],
+                           burst_std=params[f'f_stdev_{loc}'],
+                           burst_n_spikes=params[f'events_per_cycle_{loc}'],
+                           n_bursts=params[f'repeats_{loc}'])
```
Comment (jasmainak) [inline hnn_core/feed.py:42]: thinking a bit more @rythorpe I think we should just focus on this method / function and do the other methods in another PR

File: hnn_core/network.py [hunk 7]
```diff
@@ -391,6 +392,9 @@ def _create_all_spike_sources(self):
                 pc.cell(gid, self.cells[-1].connect_to_target(
                         None, self.params['threshold']))
 
+        for feed in self.feeds:
```
Comment (rythorpe) [inline hnn_core/network.py:395]: So our goal here is to set spike inputs in their own loop (i.e., separate from looping through `self._gid_list`)? Why is this necessary?

Also, `self.feeds` should be a list of both `self.feeds_common` and `self.unique_feeds` right?

File: hnn_core/feed.py (hunk 6 above)
Comment (rythorpe) [reply hnn_core/feed.py:42]: @jasmainak so just use the new `Evoked` and `Rhythmic` classes to initialize and store feed attributes and let the `Network` class handle setting connections for now?

**Comment by rythorpe** [comment]
@jasmainak Can I push something I've been working on?

**Comment by jasmainak** [comment]
Go for it!

File: hnn_core/feed.py [hunk 8]
```diff
@@ -5,219 +5,10 @@
...
-                g_basket_nmda=params['input_dist_A_weight_L2Basket_nmda'],
-                delay=params['input_dist_A_delay_L5'])
-            self.set_dist_connections_layer5(
-                g_pyr_ampa=params['input_dist_A_weight_L5Pyr_ampa'],
-                g_pyr_nmda=params['input_dist_A_weight_L5Pyr_nmda'],
-                delay=params['input_dist_A_delay_L2'])
-
-    def set_times(self, tstart, tstop, tstart_dev, burst_freq=10, burst_std=20,
-                  burst_n_spikes=2, n_bursts=10, randomize_tstart=False,
-                  random_state1=None, random_state2=None):
```
Comment (jasmainak) [inline hnn_core/feed.py]: not sure what you were attempting but I think we should keep this method as a function. Let's talk more tomorrow. Maybe I'm overlooking some challenge

File: hnn_core/network.py [hunk 9]
```diff
@@ -410,9 +410,18 @@ def _create_all_spike_sources(self):
                             params=self.p_common[p_ind],
                             gid=gid))
 
+                # Convert event times into nrn vector
```
Comment (jasmainak) [inline hnn_core/network.py:413]: I think this should be a function:

```py
def artificial_cell(event_times):
   # the code below
   # should maybe go into cell.py
   return nc

event_times = rhythmic_input_times(tstart, tstop, burst_freq, n_burst_spikes, ...)
nc = artificial_cell(event_times)
pc.cell(gid, nc)
```

File: hnn_core/network.py [hunk 10]
```diff
@@ -410,9 +412,12 @@ def _create_all_spike_sources(self):
                             params=self.p_common[p_ind],
                             gid=gid))
 
-                # create the cell and artificial NetCon
-                pc.cell(gid, self.common_feeds[-1].connect_to_target(
-                        self.params['threshold']))
+                event_times = get_rhythmic_times(tstart=params['t0'],
+                                                 tstop=params['T'],
+                                                 ...)  # TODO: fill the rest
```
Comment (jasmainak) [inline hnn_core/network.py:417]: @rythorpe I probably broke your code :) But I think I am getting a clearer and clearer picture now. I made a skeleton/sketch of what we want to accomplish in this PR. Please take a look and we'll discuss tomorrow. We really should explicitly pass parameters to make the code more transparent. If you get the idea already, feel free to pull the latest commit and continue on top of it. Let's catch up tomorrow in any case!

**Comment by rythorpe** [comment]
@jasmainak We should be clear to move forward with the PR now. At some point we should work on adding more tests for the `cell` module.

File: hnn_core/network.py [hunk 11]
```diff
@@ -155,6 +156,9 @@ def __init__(self, params, n_jobs=1):
         self._gid_assign()
         # create cells (and create self.origin in create_cells_pyr())
         self.cells = []
+        # list of artificial cells for storing external feed sources
```
Comment (jasmainak) [inline hnn_core/network.py:159]: rather than this comment, I would prefer a more informative comment which explains why you need to create the list in the first place and how it needs to be a list so that the netcon doesn't deactivate

File: hnn_core/cell.py [hunk 12]
```diff
@@ -14,6 +14,41 @@
...
 
+class ArtificialCell:
+    """The ArtificialCell class for initializing NEURON feed sources.
+    Parameters
+    ----------
+    event_times : list
+        Spike times associated with a given gid
+    threshold : float
+        The properties of the soma. Must contain
+        keys 'L', 'diam', and 'pos'
```
Comment (jasmainak) [inline hnn_core/cell.py:25]: this makes no sense :-)

File: hnn_core/cell.py [hunk 13]
```diff
@@ -14,6 +14,41 @@
 # Units for gbar: S/cm^2
 
 
+class ArtificialCell:
+    """The ArtificialCell class for initializing NEURON feed sources.
+    Parameters
+    ----------
+    event_times : list
+        Spike times associated with a given gid
```
Comment (jasmainak) [inline hnn_core/cell.py:22]: where is the gid?

File: hnn_core/network.py [hunk 14]
```diff
@@ -410,9 +412,10 @@ def _create_all_spike_sources(self):
                             params=self.p_common[p_ind],
                             gid=gid))
 
-                # create the cell and artificial NetCon
-                pc.cell(gid, self.common_feeds[-1].connect_to_target(
-                        self.params['threshold']))
+                event_times = self.common_feeds[-1].event_times
```
Comment (jasmainak) [inline hnn_core/network.py:415]: why do we still append the feeds to a list? it's not needed anymore

File: mod/vecevent.mod [hunk 15]
```diff
@@ -1,7 +1,66 @@
-:  Vector stream of events
+COMMENT
```
Comment (jasmainak) [inline mod/vecevent.mod:1]: Let's undo these changes. I copied it from PyNN, I'm not exactly sure if their license allows us to copy

File: hnn_core/feed.py [hunk 16]
```diff
@@ -5,7 +5,7 @@
 #          Christopher Bailey <bailey.cj@gmail.com>
 
 import numpy as np
-from neuron import h
+# from .utils import check_random_state
```
Comment (jasmainak) [inline hnn_core/feed.py:8]: this PR was originally meant to clean up ExtFeed object. Do you still want to do this in the PR or should we delegate this to another PR? If we do this in another PR, let's delete the above line and the corresponding function.

File: hnn_core/cell.py [hunk 17]
```diff
@@ -14,6 +14,41 @@
...
+    threshold : float
+        The properties of the soma. Must contain
+        keys 'L', 'diam', and 'pos'
+
+    Attributes
+    ----------
+    nrn_eventvec : h.Vector()
```
Comment (jasmainak) [inline hnn_core/cell.py:29]:
```suggestion
    nrn_eventvec : instance of h.Vector()
```

File: hnn_core/network.py (hunk 11 above)
Comment (rythorpe) [reply hnn_core/network.py:159]: Good point!

File: hnn_core/cell.py (hunk 13 above)
Comment (rythorpe) [reply hnn_core/cell.py:22]: The gid does not need to specified/known within the scope of this instance, however, I thought that the user should be aware that each `ArtificialCell` instance is created for a specific gid in the `Network` class.

File: hnn_core/network.py (hunk 14 above)
Comment (rythorpe) [reply hnn_core/network.py:415]: You're correct, my bad.

File: mod/vecevent.mod (hunk 15 above)
Comment (rythorpe) [reply mod/vecevent.mod:1]: So rebase?

File: hnn_core/feed.py (hunk 16 above)
Comment (rythorpe) [reply hnn_core/feed.py:8]: Yeah, let's delegate this to another PR.

**Comment by rythorpe** [comment]
@jasmainak I'm considering making the following changes:

1) interactive rebase to delete the changes to `vecevent.mod` in https://github.com/jonescompneurolab/hnn-core/pull/108/commits/fc60b7fcfcc47def543f1233152ba7d7ed5f6f29

2) deleting the `check_random_state` code in a new commit to retain history

3) additional commits to address your other concerns mentioned above

Do these sound good to you? If so, do you have any commits to push before I work on this?

**Comment by jasmainak** [comment]
@rythorpe that sounds like the right sequence of steps!

File: hnn_core/tests/test_feed.py [hunk 18]
```diff
@@ -32,41 +29,20 @@ def test_extfeed():
...
-                assert ei.params[layer + 'Pyr_ampa'][0] == params[key]
+        assert feed.feed_type == 'common'
+        assert feed.cell_type is None  # artificial cell
+        # parameters should lead to 0 input spikes for default params
+        assert len(feed.event_times) == 0
+        # check that ei.p_ext matches params
+        loc = feed.params['loc'][:4]  # loc=prox or dist
+        for layer in ['L2', 'L5']:
+            key = 'input_{}_A_weight_{}Pyr_ampa'.format(loc, layer)
+            assert feed.params[layer + 'Pyr_ampa'][0] == params[key]
```
Comment (rythorpe) [inline hnn_core/tests/test_feed.py:48]: Since I got rid of the `Network.common_feeds` list attribute, I also had to shuffle around some elements of the common feed test. I ended up moving a few elements to the `test_network.py` module. @cjayb given that you've spend a lot more time thinking about this test, did I do this correctly/adequately?

File: hnn_core/network.py [hunk 19]
```diff
@@ -423,15 +418,14 @@ def _create_all_spike_sources(self):
...
-                        self.unique_feeds[src_type][-1].connect_to_target(
-                            self.params['threshold']))
+                unique_feed = ExtFeed(feed_type=src_type,
+                                      target_cell_type=target_cell_type,
+                                      params=self.p_unique[src_type],
+                                      gid=gid)
+                self._artificial_cells.append(
+                    ArtificialCell(unique_feed.event_times,
+                                   self.params['threshold']))
+                pc.cell(gid, self._artificial_cells[-1].nrn_netcon)
```
Comment (jasmainak) [inline hnn_core/network.py:428]: can we rename `_artificial_cells` to be `_feed_cells` to make it clear that these cells relate to the external feed?

Now we have this incongruity that the class `ExtFeed` does not really create a "feed" but just gives us event_times. I guess it will be fixed once we separate the class properly

File: hnn_core/cell.py [hunk 20]
```diff
@@ -14,6 +14,42 @@
 # Units for gbar: S/cm^2
 
 
+class ArtificialCell:
```
Comment (jasmainak) [inline hnn_core/cell.py:17]: private class for now? we can expose it to user later if needed.

File: hnn_core/network.py (hunk 19 above)
Comment (rythorpe) [reply hnn_core/network.py:428]: Sure, we can do that. So maybe the feed module will eventually contain a single function, callable by the `Network` class, that creates instances of `ExtFeed` and `ArtificialCell` to be passed into `pc.cell()`?

Comment (jasmainak) [reply hnn_core/network.py]: right, but for command-line interface, I think we need to give user the ability to create feed objects outside of the `Network` class and then the `Network` class will simply create the `_feed_cells` from these objects.

File: hnn_core/tests/test_network.py [hunk 21]
```diff
@@ -35,6 +42,17 @@ def test_network():
     # Assert that an empty Spikes object is created as an attribute
     assert net.spikes == Spikes()
 
+    # Assert that all external feeds are initialized
+    net._create_all_spike_sources()
+    n_evoked_sources = 270 * 3
+    n_pois_sources = 270
+    n_gaus_sources = 270
+    n_common_sources = 2
```
Comment (jasmainak) [inline hnn_core/tests/test_network.py:50]: are these values hardcoded for the network? Or does it depend on the params?

File: hnn_core/tests/test_feed.py [hunk 22]
```diff
@@ -32,45 +29,20 @@ def test_extfeed():
...
-        net._create_all_spike_sources()
-
-        assert len(net.common_feeds) == 2  # (distal & proximal)
-        for ei in net.common_feeds:
-            assert ei.feed_type == 'common'
-            assert ei.cell_type is None  # artificial cell
-            assert hasattr(ei, 'nrn_eventvec')
-            assert hasattr(ei, 'nrn_vecstim')
-            assert ei.nrn_eventvec.hname().startswith('Vector')
-            assert hasattr(ei.nrn_vecstim, 'play')
```
Comment (jasmainak) [inline hnn_core/tests/test_feed.py:68]: @rythorpe I think you can have these tests for the new object now?

File: hnn_core/tests/test_feed.py (hunk 18 above)
Comment (jasmainak) [reply hnn_core/tests/test_feed.py:48]: do we lose some information or debuggability because of this change?

File: hnn_core/feed.py [hunk 23]
```diff
@@ -108,7 +105,7 @@ def set_event_times(self, inc_evinput=0.0):
         elif len(matches) > 1:
             raise ValueError('Ambiguous external feed: %s' % self.feed_type)
 
-        # Each of these methods creates self.nrn_eventvec for playback
+        # Each of these methods creates self.event_times for playback
```
Comment (jasmainak) [inline hnn_core/feed.py:108]:
```suggestion
        # Each of these methods creates self.event_times
```

File: hnn_core/cell.py [hunk 24]
```diff
@@ -14,6 +14,42 @@
 # Units for gbar: S/cm^2
 
 
+class _ArtificialCell:
```
Comment (jasmainak) [inline hnn_core/cell.py:17]: we need unit tests for the new object!

**Comment by jasmainak** [comment]
Otherwise LGTM. Let's try to merge this PR soonish so we can move on with life :)

File: hnn_core/tests/test_network.py (hunk 21 above)
Comment (rythorpe) [reply hnn_core/tests/test_network.py:50]: they depend on the params

File: hnn_core/tests/test_feed.py (hunk 18 above)
Comment (cjayb) [reply hnn_core/tests/test_feed.py:48]: I like these changes. Took me a while to remind myself of the challenges involved, but I think the new cell class will help in the long run. Ping self at #104

Comment (rythorpe) [reply hnn_core/tests/test_feed.py:48]: I'm also adding some unit tests for the `_ArtificialCell` class (stay tuned for an update). It incorporates most of the same elements I accidentally deleted from the next few lines of code (i.e., lines 65-68). Thanks for looking into this @cjayb!

**Comment by rythorpe** [comment]
@jasmainak check out the test I added for the `_ArtificialCell` class. If we're good there, I'm happy with merging and moving on

File: hnn_core/tests/test_cell.py [hunk 25]
```diff
@@ -16,3 +17,16 @@ def test_dipole():
     with Network(params) as net:
         net.build()
         net.cells[0].plot_voltage()
+
+
+def test_artificial_cell():
```
Comment (jasmainak) [inline hnn_core/tests/test_cell.py:22]: docstring is missing. It's displayed during test

**Comment by jasmainak** [comment]
@rythorpe please set the PR to [MRG] once it's good to go for you

**Comment by rythorpe** [comment]
> @rythorpe please set the PR to [MRG] once it's good to go for you

Just for reference, what does MRG stand for?

**Comment by jasmainak** [comment]
[MRG] = merge
[WIP] = work in progress 

these are two common titles used for PRs :) In the days when github didn't have the approve button folks would also do [MRG +1] and [MRG +2] and then merge the PR once it's [MRG +2]

**Comment by jasmainak** [comment]
whatever happened to Travis CI?

**Comment by jasmainak** [comment]
@blakecaldwell should we wait for your PR #79 to be merged first before merging this to avoid rebase mess? What do you think? One of the PRs has to do the rebase ... not sure which one is easier

**Comment by blakecaldwell** [comment]
> @blakecaldwell should we wait for your PR #79 to be merged first before merging this to avoid rebase mess? What do you think? One of the PRs has to do the rebase ... not sure which one is easier

Thanks for checking. Yeah, that would be nice to push #79 first. After these two, we will have a good handle on the `Network` and `NeuronNetwork` refactoring that remains.

**Comment by jasmainak** [comment]
@rythorpe we will need to rebase this PR now. Up for the challenge? ;-)

**Comment by rythorpe** [comment]
> @rythorpe we will need to rebase this PR now. Up for the challenge? ;-)

I need to spend some time wrapping my head around the new structure, so yeah, I'll give it a shot!

**Comment by jasmainak** [comment]
@rythorpe I did a rebase after squash. Can you check if this was done correctly? If not, can you push fixes?

**Comment by rythorpe** [comment]
Thanks for the rebase @jasmainak! Aside from the TravisCI issue on the MacOS build job, all tests now pass. I'm ready to move forward whenever you and Travis are.

**Comment by jasmainak** [comment]
okay let's move on. The MacOS build needs to be fixed separately. We should probably pin the NEURON version for now.

thanks a lot for the effort @rythorpe !
